src/box/lua/config/utils/log.lua

```diff
@@ -0,0 +1,135 @@
+-- Logger wrapper with a few enhancements.
+--
+-- 1. Encode tables into JSON.
+-- 2. Enable all the messages on TT_CONFIG_DEBUG=1.
+--
+-- Hopefully, the wrapper can be dropped when the following logger
+-- problems will be solved.
+--
+-- * JSONify tables at logging (gh-8611)
+-- * Follow TT_LOG_LEVEL before log.cfg()/box.cfg() (gh-8092)
+--
+-- A solution of the following problems may simplify the code of
+-- this wrapper.
+--
+-- * mylog.cfg() configures the main logger (gh-8610)
+-- * No easy way to determine if a message should go to the log
+--   (gh-8730)
+
+local json_noexc = require('json').new()
+json_noexc.cfg({encode_use_tostring = true})
+
+local logger_name = 'tarantool.config'
+local log = require('log').new(logger_name)
+
+local func2level = {
+    [log.error] = 2,
+    [log.warn] = 3,
+    [log.info] = 5,
+    [log.verbose] = 6,
+    [log.debug] = 7,
+}
+
+local func2prefix = {
+    [log.error] = 'E> ',
+    [log.warn] = 'W> ',
+    [log.info] = 'I> ',
+    [log.verbose] = 'V> ',
+    [log.debug] = 'D> ',
+}
+
+local str2level = {
+    ['fatal'] = 0,
+    ['syserror'] = 1,
+    ['error'] = 2,
+    ['warn'] = 3,
+    ['crit'] = 4,
+    ['info'] = 5,
+    ['verbose'] = 6,
+    ['debug'] = 7,
+}
+
+local function say_closure(log_f)
+    local prefix = ''
+
+    -- Enable logging of everything if an environment variable is
+    -- set.
+    --
+    -- Useful for debugging.
+    --
+    -- Just setting of...
+    --
+    -- ```
+    -- TT_LOG_MODULES='{"tarantool.config": "debug"}'
+    -- ```
+    --
+    -- ...is not suitable due to gh-8092: messages before first
+    -- box.cfg() are not shown.
+    --
+    -- Explicit calling of...
+    --
+    -- ```
+    -- log.cfg({modules = {['tarantool.config'] = 'debug'}})
+    -- ```
+    --
+    -- is not suitable as well, because it makes the logger
+    -- already configured and non-dynamic options like
+    -- `box_cfg.log_nonblock` can't be applied at first box.cfg()
+    -- invocation.
+    --
+    -- So just prefix our log messages and use log.info().
+    local envvar = os.getenv('TT_CONFIG_DEBUG')
+    if envvar ~= nil and envvar ~= '0' and envvar ~= 'false' then
+        prefix = func2prefix[log_f]
+        log_f = log.info
+    end
+
+    return function(fmt, ...)
+        -- Skip logging based on the log level before performing
+        -- the encoding into JSON.
+        --
+        -- log.new(<...>).cfg is the main logger configuration,
+        -- the same as log.cfg. Extract the current log level from
+        -- log.cfg.modules if it is set specifically for the config
+        -- module. See gh-8610.
+        local level = log.cfg.modules and log.cfg.modules[logger_name] and
+            log.cfg.modules[logger_name] or log.cfg.level
```

Comment:
I'll play with the code and return here back.

Reply:
Let's compare.

```lua
local level = log.cfg.modules and log.cfg.modules[logger_name] or
    log.cfg.level
```

```lua
local level
if log.cfg.modules ~= nil and log.cfg.modules[logger_name] ~= nil then
    level = log.cfg.modules[logger_name]
else
    level = log.cfg.level
end
```

```lua
local level = log.cfg.level
if log.cfg.modules ~= nil and log.cfg.modules[logger_name] ~= nil then
    level = log.cfg.modules[logger_name]
end
```

```lua
local level = log.cfg.modules and log.cfg.modules[logger_name]
if not level then
    level = log.cfg.level
end
```

```lua
local level = log.cfg.modules and log.cfg.modules[logger_name]
level = level or log.cfg.level
```

Or you're just about line carrying?

Instead of

```lua
local level = log.cfg.modules and log.cfg.modules[logger_name] or
    log.cfg.level
```

write it like so?

```lua
local level = log.cfg.modules and
    log.cfg.modules[logger_name] or
    log.cfg.level
```

or so?

```lua
local level = log.cfg.modules
    and log.cfg.modules[logger_name]
    or log.cfg.level
```

It would look like a ternary operator, but it doesn't work this way. Not `x ? a : b`, but `x ? (a ? a : b) : b`.

I would stick with the following.

```lua
local level = log.cfg.modules and log.cfg.modules[logger_name] or
    log.cfg.level
```

Reply:
I'm about this variant:

```lua
local level = log.cfg.modules and
    log.cfg.modules[logger_name] or
    log.cfg.level
```

for my taste, it is more readable than others.

In this comment we have both opposite opinion again, let's wait an opinion from a third reviewer.

Reply:
I'd rather write this the following way:
```lua
local level = log.cfg.modules
    and log.cfg.modules[logger_name]
    or log.cfg.level
```

However, our style guide forbids it (IIRC). I personally prefer @ligurio way, but the current code is also fine, so, @Totktonada, feel free to choose any of them since you're the author of these lines.

Reply:
No problem. Updated:

```diff
diff --git a/src/box/lua/config/utils/log.lua b/src/box/lua/config/utils/log.lua
index 1cfb991d23..3072c77ddc 100644
--- a/src/box/lua/config/utils/log.lua
+++ b/src/box/lua/config/utils/log.lua
@@ -95,7 +95,8 @@ local function say_closure(log_f)
         -- the same as log.cfg. Extract the current log level from
         -- log.cfg.modules if it is set specifically for the config
         -- module. See gh-8610.
-        local level = log.cfg.modules and log.cfg.modules[logger_name] or
+        local level = log.cfg.modules and
+            log.cfg.modules[logger_name] or
             log.cfg.level
         -- The level is either a string or a number. Transform it
         -- to a number. See gh-8730.
```

Reply:
Thanks! Resolving then.